sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala

```diff
@@ -0,0 +1,168 @@
+ *
+ * Note: [[HDFSMetadataLog]] doesn't support S3-like file systems as they don't guarantee listing
+ * files in a directory always shows the latest files.
+ */
+class HDFSMetadataLog[T: ClassTag](sqlContext: SQLContext, path: String) extends MetadataLog[T] {
+
+  private val metadataPath = new Path(path)
+  private val fs = metadataPath.getFileSystem(sqlContext.sparkContext.hadoopConfiguration)
+  if (!fs.exists(metadataPath)) {
+    fs.mkdirs(metadataPath)
+  }
+
+  tryAcquireLock()
+
+  /**
+   * A `PathFilter` to filter only batch files
+   */
+  private val batchFilesFilter = new PathFilter {
+    override def accept(path: Path): Boolean = try {
+      path.getName.toLong
+      true
+    } catch {
+      case _: NumberFormatException => false
+    }
+  }
+
+  private val serializer = new JavaSerializer(sqlContext.sparkContext.conf).newInstance()
+
+  private def tryAcquireLock(): Unit = {
+    val lockFile = new Path(metadataPath, ".lock")
+    fs.deleteOnExit(lockFile)
+    try {
+      // Must set overwrite to false otherwise HDFS allows multiple clients to create the same file.
+      // When overwrite is false, the latter one will receive FileAlreadyExistsException
+      fs.create(lockFile, false).close()
+    } catch {
+      case e: IOException =>
+        throw new SparkException(s"There are another user using $path. Please check that. " +
+          s"If nobody is using it, please delete ${lockFile}", e)
+    }
+  }
+
+  private def batchFile(batchId: Long): Path = {
+    new Path(metadataPath, batchId.toString)
+  }
+
+  override def add(batchId: Long, metadata: T): Unit = {
+    get(batchId).getOrElse {
+      // Only write metadata when the batch has not yet been written.
+      val buffer = serializer.serialize(metadata)
+      val output = try {
+        fs.create(batchFile(batchId))
+      } catch {
+        case e: IOException if "java.lang.InterruptedException" == e.getMessage =>
+          // create may convert InterruptedException to IOException. Let's convert it back to
+          // InterruptedException so that this failure won't crash StreamExecution
+          throw new InterruptedException("Creating file is interrupted")
+      }
+      try {
+        output.writeInt(buffer.remaining())
+        Utils.writeByteBuffer(buffer, output: DataOutput)
+      } finally {
+        output.close()
+      }
+    }
+  }
+
+  override def get(batchId: Long): Option[T] = {
+    val batchMetadataFile = batchFile(batchId)
+    if (fs.exists(batchMetadataFile)) {
+      val input = fs.open(batchMetadataFile)
+      try {
+        val size = input.readInt()
+        val bytes = new Array[Byte](size)
+        val readSize = input.read(bytes)
+        if (readSize == size) {
+          Some(serializer.deserialize[T](ByteBuffer.wrap(bytes)))
+        } else {
+          None
+        }
+      } catch {
+        case _: EOFException =>
+          // The file is corrupted, so return None
```

Comment:
I'm not sure if this is a big deal, but this does not handle the case where there is garbage at the end of the file, right?